joomla/joomla-cms
# [4.0] Improve CSP, second attempt

### Summary of Changes

It seems most confusing part of CSP is "mode" selection, especially when it set to "auto".
So I have dropped mode selection, and "auto-header" generation  (we can review it in future).
And turned the CSP component to User configurable.

Also Reporting configuration was changed. Now it possible to enable reporting and "report-only" in separated options.
This allows to collect Reports for live site with enabled "Content-Security-Policy" header.

**New CSP workflow:**
- Enable CSP component, this also will enable reporting.
- Set up CSP rules (in component config).
- Navigate through the site, visit different pages.
- Go to CSP reports page in the backed, and check if here any report.
- Adjust CSP rules  (in component config), or fix blocked content, and clean up the report list.
- (optionally) turn of the "Report"



### Testing Instructions

**Before you begin: make sure your test site domain not `localhost` or IP**

Apply patch.
Enable CSP in component configuration.
Make sure "Client" is Site
Make sure "Report" is Enabled.
Make sure "Report-Only" is Off.

**step 0**
Visit to the site, make sure it is working, nothing broken.
And the CSP report is empty.

**step 1**

Add 2 rules in CSP component configuration.
```
default-src: 'self'
img-src: 'self'
```

**step 2**
Create a custom HTML module, with background, with some text and with an external IMG:
```
<img src="https://picsum.photos/300/300" width="300" height="300" />
```

Check the module on the site.
The module will not show images.


**step 3**

Got to the CSP report page, in backend.
`administrator/index.php?option=com_csp&view=reports`

You should see at least 2 blocked elements: `'unsafe-inline'` and `https://picsum.photos/300/300`

**step 4**

Update CSP rules in the component configuration, to next:
```
default-src: 'self' 'unsafe-inline'
img-src: 'self' picsum.photos *.picsum.photos
```


Visit to the site and check that the module is working.
Clean up the report and make sure there no more blocked elements: `'unsafe-inline'` and `https://picsum.photos/300/300` after further Site navigation.


### Documentation Changes Required

Yes, the CSP documentation page need to be updated, to reflect the changes.
Additionally need to warn User that the editor may add inline style so he/she may need to add `'unsafe-inline'`. 
And that CSP and  `localhost` not a best friends.


@zero-24  please review
